lib/spack/spack/build_environment.py

```diff
@@ -287,10 +287,10 @@ def set_build_environment_variables(pkg, env, dirty):
         except spack.spec.NoLibrariesError:
             tty.debug("No libraries found for {0}".format(dep.name))
 
-        # TODO: fix the line below, currently the logic is broken for
-        # TODO: packages that uses directories as namespaces e.g.
-        # TODO: #include <boost/xxx.hpp>
-        # include_dirs.extend(query.headers.directories)
+        try:
+            include_dirs.extend(query.headers.directories)
+        except spack.spec.NoHeadersError:
+            tty.debug("No headers found for {0}".format(dep.name))
 
         if os.path.isdir(dep.prefix.include):
             include_dirs.append(dep.prefix.include)
```

Comment:
@scheibelp Is there a reason why we add `prefix.include` explicitly, and have it tested too? Asking because I don't think this is necessary anymore with the fixed logic in `HeaderList.directories`.

Reply:
See https://github.com/spack/spack/pull/8136#issuecomment-391752739

Your changes here would address that though, and thereby make this unnecessary

Reply:
@alalazo If this line stays, for similar reasons to https://github.com/spack/spack/pull/10628 please move this under `except` path so that we don't add it unconditionally.

Reply:
This line has been removed. Tests stressing directory search have been added. I'll probably add a few more unit tests involving only `HeaderList` instances.